conn/physic/units.go

```diff
@@ -733,3 +734,250 @@ func picoAsString(v int64) string {
+	if d.exp+scale < 0 {
+		u /= powerOf10[mag]
+	} else {
+		check := u * powerOf10[mag]
+		if check < u || check > max {
+			return max, &parseError{err: errors.New("overflows maximum is"), s: strconv.FormatUint(max, 10)}
+		}
+		u *= powerOf10[mag]
+	}
+
+	n := int64(u)
+	if d.neg {
+		n *= -1
+	}
+	return n, nil
+}
+
+// Converts a string to a decimal form. The return int is how many bytes of the
+// string are numeric. The string may contain +-0 prefixes and arbitrary
+// suffixes as trailing non number characters are ignored.
+// Significant digits are stored without leading or trailing zeros, rather an
+// exponet is used.
+func atod(s string) (decimal, int, error) {
+	var d decimal
+	start := 0
+	dp := 0
+	end := 0
+	seenDigit := false
+	seenZero := false
+	isPoint := false
+
+	// Strip leading zeros, +/- and mark DP.
+	for i := 0; i < len(s); i++ {
+		switch {
+		case s[i] == '-':
+			d.neg = true
+			start++
+		case s[i] == '+':
+			start++
+		case s[i] == '.':
+			if isPoint {
+				return d, 0, &parseError{s, i, errors.New("multiple decimal points")}
+			}
+			isPoint = true
+			dp = i
+			if !seenDigit {
+				start++
+			}
+		case s[i] == '0':
+			if !seenDigit {
+				start++
+			}
+			seenZero = true
+		case s[i] >= '1' && s[i] <= '9':
+			seenDigit = true
+		default:
+			if !seenDigit && !seenZero {
+				return d, 0, &parseError{s: s, err: errors.New("is not a number")}
+			}
+			end = i
+			break
+		}
+	}
+
+	if end == 0 {
+		end = len(s)
+	}
+	last := end
+	seenDigit = false
+	exp := 0
+	// Strip non significant zeros to find base exponent.
+	for i := end - 1; i > start-1; i-- {
+		switch {
+		case s[i] >= '1' && s[i] <= '9':
+			seenDigit = true
+		case s[i] == '.':
+			if !seenDigit {
+				end--
+			}
+		case s[i] == '0':
+			if !seenDigit {
+				if i > dp {
+					end--
+				}
+				if i <= dp || dp == 0 {
+					exp++
+				}
+			}
+		default:
+			last--
+			end--
+		}
+	}
+
+	if dp > start && dp < end {
+		// Concatenate with out decimal point.
+		d.digits = s[start:dp] + s[dp+1:end]
+	} else {
+		d.digits = s[start:end]
+	}
+	if !isPoint {
+		d.exp = exp
+	} else {
+		ttl := dp - start
+		length := len(d.digits)
+		if ttl > 0 {
+			d.exp = ttl - length
+		} else {
+			d.exp = ttl - length + 1
+		}
+	}
+	return d, last, nil
+}
+
+type parseError struct {
+	s        string
+	position int
+	err      error
+}
+
+func (p *parseError) Error() string {
+	if p.err == nil {
+		return "parse error"
+	}
+	return "parse error: " + p.err.Error() + ": \"" + p.s + "\""
+}
+
+func noUnits(s string) error {
+	return &parseError{s: s, err: errors.New("no units provided, need")}
+}
+
+type prefix int
+
+const (
+	pico  prefix = -12
+	nano  prefix = -9
+	micro prefix = -6
+	milli prefix = -3
+	none  prefix = 0
+	deca  prefix = 1
+	hecto prefix = 2
+	kilo  prefix = 3
+	mega  prefix = 6
+	giga  prefix = 9
+	tera  prefix = 12
+)
+
+func parseSIPrefix(r rune) (prefix, int) {
+	switch r {
+	case 'p':
+		return pico, len(string(r))
```

Comment:
Because you are using r here, I think the compile will not be able to inline this into a constant. So while less safe, I'd prefer len("p"), etc, for performance.

Reply:
That is better idea.